File: src/main/java/victor/training/cleancode/idee1_paramObj/BooleanParameters.java
package victor.training.cleancode.idee1_paramObj;

import lombok.Value;

import java.util.ArrayList;
import java.util.List;

/**
 * Implement the variation required for CR323 without adding a boolean parameter
 */
public class BooleanParameters {

   public void bossLevel(BossParamsObject params) {
      int index = 0; // TODO move closer to usages
      int j = params.getTasks().size();
      System.out.println("Logic1");
      List<Integer> taskIds = new ArrayList<>();
      System.out.println("Logic2");
      if (params.isFluff()) {
         fluffBranch(params, index, j, taskIds);
      } else {
         System.out.println("Logic7 " + params.getTasks());
      }
      System.out.println("Logic7");
   }

   private void fluffBranch(BossParamsObject params, int index, int j, List<Integer> taskIds) {
      System.out.println("Logic3");
      for (Task task : params.getTasks()) {
         System.out.println("Logic4: Validate " + task);
         task.setRunning();

         taskIds.add(task.getId());

         if (params.isCr323()) { // TODO remove the boolean
            System.out.println("My Logic: " + task);
         }

         index++;
         System.out.println("Logic5 index=" + index + " on running=" + task.isRunning());
      }
      System.out.println("Logic6 " + j);
      System.out.println("Task Ids: " + taskIds);
   }

}

// PROBLEM:  this "Parameter Object" is only used 100% for one method invocation.
// when calling fluffBranch(), only 2 fields out of 3 are really needed.
// If it was used 100% for many method calls, it might be a good idea!
@Value
class BossParamsObject {
   boolean fluff;
   List<Task> tasks;
   boolean cr323;
}


class Task {
   private final int id;
   private boolean running;

   Task(int id) {
      this.id = id;
   }

   public void setRunning() {
      running = true;
   }

   public boolean isRunning() {
      return running;
   }

   public int getId() {
      return id;
   }

   @Override
   public String toString() {
      return "Task{" + "id=" + id + ", running=" + running + '}';
   }
}

File: src/main/java/victor/training/cleancode/BooleanParameters.kt
package victor.training.cleancode

/**
 * Implement the variation required for CR323 without adding a boolean parameter
 */
class BooleanParameters {
    // ============== "BOSS" LEVEL: Deeply nested functions are a lot harder to break down =================
    // Lord gave us tests!
    fun bossLevelFluff(tasks: List<Task>, cr323: Boolean) {
        bossStart(tasks)

        if (cr323) { // TODO remove the boolean
            for (task in tasks) {
                println("My Logic: $task")
            }
        }

        bossEnd(tasks)
    }

    private fun bossEnd(tasks: List<Task>) {
        tasks.forEachIndexed {index,task->println("Audit task index=$index: $task") }
        println("Logic6 ${tasks.size}")
        val taskIds = tasks.map { it.id }
        println("Task Ids: $taskIds")
        println("Logic7")
    }

    private fun bossStart(tasks: List<Task>) {
        println("Logic1")
        println("Logic2")
        println("Logic3")
        for (task in tasks) {
            println("Logic4: Validate $task")
            task.started = true
        }
    }

    fun bossLevelNoFluff(tasks: List<Task>) {
        println("Logic1")
        println("Logic2")
        println("Logic7 $tasks")
        println("Logic7")
    }

}

class Task(val id: Int, var started: Boolean = false) {

    override fun toString(): String {
        return "Task{" + "id=" + id + ", started=" + started + '}'
    }
}